# Logos: OPATHY, CRG, BIOTECHVANA


# crossmapper

Crossmapper is an automated bioinformatics pipeline for asessing the rate of read crossmapping when two or more organisms are sequenced as one sample. The software can be used for planning such experimental set-ups as dual- or multiple RNA-seq (mainly for host-pathogen, symbiont and cohabitant interation studies), metagenomics studies, allele-specific expression studies, and can be extended for the use in large seqeuncing facilities for resource optimization.

Based on in-silico read simulation and back-mapping to the original genomes/transcirptomes of sequenced organisms, Crossmapper allows the user to assess the rate of errorneous multi- and crossmapping of reads to non-corresponding genomes and thus helps to optimise the sequencing parameters (such as read lenght, paired-end, single-end, mapping parameters, etc) to minimize the crossmapping rate.


General description, authors, citation, licence


# Step by step tutorial
In our step-by-step tutorial 

# Instalation and setup
 We have implemented Crossmap in python 3 as an Anaconda package. Thus, the user will just need to install Anaconda, which will efficiently take care of all dependancies for Crossmap.

# Contact and reporting
